pkg/meta/meta.go

```diff
@@ -0,0 +1,168 @@
+/*
+Copyright 2018 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package meta contains functions for dealing with Kubernetes object metadata.
+package meta
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+/*
+	Prefer taking *corev1.ObjectReference as an argument where possible when
+	adding new functions to this package. It's easier to convert an object to
+	an ObjectReference using ReferenceTo() than it is to make an ObjectReference
+	satisfy metav1.Object.
+*/
+
+// ReferenceTo returns an object reference to the supplied object, presumed to
+// be of the supplied group, version, and kind.
+func ReferenceTo(o metav1.Object, of schema.GroupVersionKind) *corev1.ObjectReference {
+	v, k := of.ToAPIVersionAndKind()
+	return &corev1.ObjectReference{
+		APIVersion: v,
+		Kind:       k,
+		Namespace:  o.GetNamespace(),
+		Name:       o.GetName(),
+		UID:        o.GetUID(),
+	}
+}
+
+// AsOwner converts the supplied object reference to an owner reference.
+func AsOwner(r *corev1.ObjectReference) metav1.OwnerReference {
+	return metav1.OwnerReference{
+		APIVersion: r.APIVersion,
+		Kind:       r.Kind,
```

Comment:
This could be problematic. I don't think API always sets  Kind and Type. 
Did we verify that this works?

Reply:
This takes a `*corev1.ObjectReference`, which is presumed to have type metadata. It's up to the caller to ensure the provided reference has metadata. I imagine the typical call pattern will be `meta.AsOwner(meta.ReferenceTo(object, gvk))`, in which `ReferenceTo` is responsible for generating a fully populated `*corev1.ObjectReference`.

Per the comment at the top of this file, the API is designed this way because in many cases we want to create owner references, controller references, etc from existing class, claim, or resource references rather than from a type that satisfies `metav1.Object`.